### PR TITLE
Several changes to improve Socket stress

### PR DESCRIPTION
**Pull request by geoffkizer**
Existing Socket stress test ignores most failures, such as connection failure, send failure, receive failure, etc.  I am adding checks for these failures.

When adding these checks, I noticed several other problems:
(1) On Linux, all connections were silently failing due to the way we were trying to connect.  The server was listening on IPv6Loopback, but the client was trying to connect via DNS resolution to "localhost".  On Windows, "localhost" resolves to both IPv6Loopback and IPv4Loopback, so this works.  On Linux, it only resolves to IPv4Loopback, thus all client connections fail.

Fix is to make the client explicitly connect to IPv6Loopback instead of using the "localhost" name.

(2) On Windows, the host OS would start rejecting connections at around the 150th connection or so.  This seemed to be due to the way the test works; it was connecting and immediately performing lots and lots of sends to the server.

The fix is to rework the test to do sends and receives in a slightly less aggressive way, and make sure that the client is processing receives at the same time as sends by spawning separate tasks for them.

Also, a couple related changes:
(1) Randomly vary the size of the amount of data sent before the client stops sending and waits for all data to be received from the server.
(2) Add the SOCKETSTRESS_ITERATIONS env variable to enable longer stress runs.
(3) Rework some code to use Tasks, to make the code easier to implement
(4) Have just one stress test instead of multiple.  (The second variation wasn't really providing any additional coverage anyway.)

This stress test should take around 30s to run, compared to 3s previously.

### CODE REVIEW

**Comment by geoffkizer** [comment]
@stephentoub @davidsh @CIPop @Priya91

File: src/Common/tests/System/Net/Sockets/Performance/SocketTestClient.cs [hunk 1]
```diff
@@ -104,167 +93,165 @@ public abstract class SocketTestClient
...
-            {
-                OnReceive(bytesTransferred, socketError);
+                // 
+                // TODO: It might be more efficient to have more than one outstanding Send/Receive.
+
+                // IMPORTANT: The code currently assumes one outstanding Send and one Receive. Interlocked operations
+                //            are required to handle re-entrancy.
+
+                Task t1 = Task.Run(async () => { await DoSend(); });
+                Task t2 = Task.Run(async () => { await DoReceive(); });
```
Comment (stephentoub) [inline src/Common/tests/System/Net/Sockets/Performance/SocketTestClient.cs]: Nit: the async/await is unnecessary here, could just be:
```C#
Task t1 = Task.Run(() => DoSend());
Task t2 = Task.Run(() => DoReceive());
```

Comment (geoffkizer) [inline src/Common/tests/System/Net/Sockets/Performance/SocketTestClient.cs]: Fixed

**Comment by geoffkizer** [comment]
@dotnet-bot 
test outerloop OSX Debug 
test outerloop OSX Release
test outerloop Ubuntu14.04 Debug 
test outerloop Ubuntu14.04 Release
test outerloop Windows_NT Debug
test outerloop Windows_NT Release

**Comment by geoffkizer** [comment]
Even with these changes, we were seeing some ConnectionRefused errors on Windows_NT debug.

To fix this, add some connection retry logic into the client.

**Comment by geoffkizer** [comment]
@dotnet-bot
test outerloop OSX Debug
test outerloop OSX Release
test outerloop Ubuntu14.04 Debug
test outerloop Ubuntu14.04 Release
test outerloop Windows_NT Debug
test outerloop Windows_NT Release

**Pull request merged**